boards/arm64/mimx93_evk/mimx93_evk_a55_sof.dts

```diff
@@ -85,4 +85,7 @@
 	/* clocks = <&ccm IMX_CCM_UART4_CLK 0x6c 24>; */
 	pinctrl-0 = <&uart2_default>;
 	pinctrl-names = "default";
+	assigned-clocks = <IMX93_CCM_LPUART2_ROOT>;
+	assigned-clock-parents = <IMX93_CCM_OSC_24M>;
```

Comment:
It seems like what we are trying to do here is encode a clock tree with these two properties. Rather than having a `clock` and `clock-parent` property, why not describe the `clock` node as a devicetree node, then have the parent of that node be the source clock?

Reply:
> It seems like what we are trying to do here is encode a clock tree with these two properties. Rather than having a `clock` and `clock-parent` property, why not describe the `clock` node as a devicetree node, then have the parent of that node be the source clock?

I think we are trying to follow here the Linux implementation which is proved to work for a while.

Reply:
Ok, I've taken a look at the Linux implementation, and I think I understand the goals here a bit better now. With that being said, Linux clearly has different clock drivers for each SOC (since the clock driver itself encodes the clock tree data, and as far as I can see Linux provides a generic framework to handle a lot of common functionalities with clocks: https://docs.kernel.org/driver-api/clk.html.)

Given that, wouldn't we want this PR to have two components:
- a common clock framework of some form (that we can initially use with NXP clock drivers, then attempt to expand to be a generic functionality)
- an SOC specific driver for the iMX 93 part

Reply:
Additionally- just a general question about Linux clock setup- how does Linux handle a case where the initial clock frequency of the SOC should be produced via a PLL (something like 24MHz OSC -> PLL0 -> output clock)? Is the idea that the clock framework would call `set_rate` for the PLL, and then the PLL divider and multiplier would be calculated at runtime?

My concern with this approach in general is that for larger SOCs like the iMX RT series (or iMX itself), this approach makes sense for the ease of use. However, on smaller cores like the LPC parts I am concerned about the overhead of performing these clock calculations on boot

Reply:
> Ok, I've taken a look at the Linux implementation, and I think I understand the goals here a bit better now. With that being said, Linux clearly has different clock drivers for each SOC (since the clock driver itself encodes the clock tree data, and as far as I can see Linux provides a generic framework to handle a lot of common functionalities with clocks: https://docs.kernel.org/driver-api/clk.html.)
> 
> Given that, wouldn't we want this PR to have two components:
> 
> * a common clock framework of some form (that we can initially use with NXP clock drivers, then attempt to expand to be a generic functionality)
> * an SOC specific driver for the iMX 93 part

It does. `clock_control_mcux_ccm_rev3.c` is supposed to contain the SoC-independent operations such as parsing the clock tree to set a clock's rate, managing the software states of the clocks (`GATED`, `UNGATED`) etc....(basically the common clock framework). The SoC specific driver is in `imx93_ccm.c` which implements the SoC-specific operations such as setting a clock's rate, setting a clock's parent etc...


> Additionally- just a general question about Linux clock setup- how does Linux handle a case where the initial clock frequency of the SOC should be produced via a PLL (something like 24MHz OSC -> PLL0 -> output clock)? Is the idea that the clock framework would call `set_rate` for the PLL, and then the PLL divider and multiplier would be calculated at runtime?

Not sure how helpful I can be here since I'm not entirely familiar with how Linux initializes the clocks (also this approach isn't meant to be a 1:1 recreation of the Linux CCF, but rather a very simplified version that would still be powerful enough to serve our purposes). Not sure if this would help but the way I see things there's 3 possible approaches:

1. During initialization every clock is set to a default configuration which can then be altered later on using the `assigned-clock*` properties.
2. During initialization only clocks configured through `assigned-clock*` properties are made usable (also the clocks which would be configured through propagation). This would leave unconfigured clocks basically unusable so the driver should raise an error when attempting to use a clock which depends on one of these clocks.
3. The SoC layer itself knows the default configuration of the clocks so it just informs the upper layer such that no initialization is required.

I tend to think option 1 (or even 3 maybe?) would be more plausible in Linux's case. As a side note, my current approach is based on number 2. Ideally, it should be based on 1 or 3.

> My concern with this approach in general is that for larger SOCs like the iMX RT series (or iMX itself), this approach makes sense for the ease of use. However, on smaller cores like the LPC parts I am concerned about the overhead of performing these clock calculations on boot

Yeah this is indeed concerning. We could optimize the driver such that it initializes only clocks used by consumers, instead of initializing all of the clocks declared in the SoC layer's clock tree. Other than that, you could reduce the overhead by using a toned-down version of the clock tree and reducing the amount of configurations that need to be done through `assigned-clock*` properties.

Reply:
> It does. clock_control_mcux_ccm_rev3.c is supposed to contain the SoC-independent operations such as parsing the clock tree to set a clock's rate, managing the software states of the clocks (GATED, UNGATED) etc....(basically the common clock framework). The SoC specific driver is in imx93_ccm.c which implements the SoC-specific operations such as setting a clock's rate, setting a clock's parent etc...

This makes sense- thank you for the clarification

> 1. During initialization every clock is set to a default configuration which can then be altered later on using the assigned-clock* properties.
2. During initialization only clocks configured through assigned-clock* properties are made usable (also the clocks which would be configured through propagation). This would leave unconfigured clocks basically unusable so the driver should raise an error when attempting to use a clock which depends on one of these clocks.
3. The SoC layer itself knows the default configuration of the clocks so it just informs the upper layer such that no initialization is required.

Of these, I actually prefer 2), since that would result in the lowest runtime overhead for the platform, and the least amount of definitions needed in code.

The key thing I am worried about here is a need to implement `set_rate` on all clock devices. For example, if we define the clock tree itself in the DTS, we could end up with something like this for a PLL:

```
osc_24m {
	clock-frequency = <24000000>;
};

rtc_32k {
	clock-frequency = <32768>;
};


pll_clk_sel {
	input-sources = <&osc_24m &rtc_32k>;
	pll0: pll0 {
		mult = <25>;
		div = <1>;
	};
};
```

with something like this, the SOC initialization code could easily determine the PLL settings needed. If we used `set_rate`, then the SOC code would need to calculate the PLL multiplier and divider at runtime. The other issue I am concerned about is code size. With a devicetree, the C code will come down to a function call (and could be as simple as a few register writes on some platforms). With `set_rate` implementations, the code will need to include logic to perform these calculations at runtime